Finished Chapter 3 solutions

jacobprudhomme opened the pull request:
### Solutions for Chapter 3

For now, I am stuck on the boss task. Do you have any hints for how to proceed? :smiley: 

I am moving in the direction of having a typeclass `Fighter`, of which I will make an instance for every type that can engage in a fight. This typeclass will have a method `fight'`, which is passed the attacker, the defender, and the string of the action to take (`"attack"`, `"run"`, `"defend"`, etc.). It then returns a tuple of the attacker and defender, updated to have their stat changes reflected (higher defense, lower health, etc.).

However, the way I have it currently written out, I keep getting type errors like the following (note that I did not commit all my work):
```
• Couldn't match expected type ‘b’ with actual type ‘Monster'’
  ‘b’ is a rigid type variable bound by
    the type signature for:
      fight' :: forall b. Monster' -> b -> String -> (Monster', b)
    at /home/jacobprudhomme/Workspace/learn4haskell/src/Chapter3.hs:1153:13-52
• In the pattern: Monster' mh2 ma2
  In an equation for ‘fight'’:
      fight' (Monster' mh1 ma1) (Monster' mh2 ma2) "attack"
        = let
            attack = getAttack' (monsterAttack' ma1)
            health = getHealth' (monsterHealth' mh2)
            fighter = Monster' mh1 ma1
            opponent = Monster' (Health' (health - attack)) ma2
          in (fighter, opponent)
  In the instance declaration for ‘Fighter Monster'’
• Relevant bindings include
    fight' :: Monster' -> b -> String -> (Monster', b)
      (bound at /home/jacobprudhomme/Workspace/learn4haskell/src/Chapter3.hs:1154:3)
```
I am not sure how to proceed, or if I should completely change my design.

Any help is much appreciated!

cc @vrom911 @chshersh

jacobprudhomme committed:
Finished Chapter 3 solutions (only partially finished boss task)

chshersh commented on src/Chapter3.hs:
```diff
@@ -343,6 +343,15 @@ Define the Book product data type. You can take inspiration from our description
 of a book, but you are not limited only by the book properties we described.
 Create your own book type of your dreams!
 -}
+data Book = MkBook
+  { bookName :: String
+  , bookAuthor :: String
+  , bookCover :: FilePath
+  , bookPages :: Int
+  , bookGenres :: [String]
+  , bookPublisher :: String
+  , bookYearPublished :: Int
+  }
```
Very creative fields! :books:

chshersh commented on src/Chapter3.hs:
```diff
@@ -459,6 +485,19 @@ and provide more flexibility when working with data types.
 Create a simple enumeration for the meal types (e.g. breakfast). The one who
 comes up with the most number of names wins the challenge. Use your creativity!
 -}
+data Meal
+  = PreSwimmingLesson5AMSnack
+  | LittleBreakfast
+  | Breakfast
+  | MidMorningSnack
+  | Lunch
+  | ExtendedLunch
+  | MidAfternoonSnack
+  | DinnerAppetizer
+  | DinnerSoup
+  | DinnerEntree
+  | DinnerDessert
+  | MidnightSnack
```
Very nice menu! :heart_eyes: I see a candidate for winning :1st_place_medal:

jacobprudhomme replied:
Thanks, I tried my best!

chshersh commented on src/Chapter3.hs:
```diff
@@ -479,6 +518,37 @@ After defining the city, implement the following functions:
+
+data Wall = Wall | NoWall
+
+data EducationCentre = Church | Library
```
:rofl:

jacobprudhomme replied:
I hoped somebody would get the joke :wink:

chshersh commented on src/Chapter3.hs:
```diff
@@ -479,6 +518,37 @@ After defining the city, implement the following functions:
+
+data House = OnePerson | TwoPerson | ThreePerson | FourPerson
+
+data City
+  = CityWithCastle Castle Wall EducationCentre [House]
+  | CityWithoutCastle EducationCentre [House]
```
Although this totally makes sense, it would be much easier to have `City` as a product type.
Let's see how it fits:

first, we need to decide with the castle. Let's look at the task:

> Wall, but only if the city has a castle
> -- Optional castle with a name (as 'String')

That means that walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

 1. Castle with a name
 2. Castle with the name + walls
 3. No castle

This looks like a sum type!:

```haskell
data Castle 
    = Castle String
    | WallCastle String
    | None
```

Then let's build the city:

```haskell
data City = City
    -- Optional castle with a __name__ (as 'String')
    -- Wall, but only if the city has a castle
    { cityCastle :: Castle -- from the above
    -- Church or library but not both
    , cityChurchOrLibrary :: EducationCentre -- you have already created this one
    -- any number of houses. Each house has one, two, three or four __people__ inside.
    , cityHouses :: [House]
    }
```

And after that, you would be able to implement other functions easily. Let me know if this helps!

chshersh commented on src/Chapter3.hs:
```diff
@@ -560,22 +630,30 @@ introducing extra newtypes.
-calculatePlayerHit :: Int -> Int -> Int -> Int
-calculatePlayerHit damage defense health = health + defense - damage
+calculatePlayerHit :: Damage -> Defense -> Health -> Health
+calculatePlayerHit dmg def hlt = Health (getHealth hlt + getDefense def - getDamage dmg)
```
Nice work on newtypes! :medal_sports: Newtypes are one very important feature in Haskell :+1:
You can read more about the newtype pattern in our blog post: https://kowainik.github.io/posts/haskell-mini-patterns#newtype

chshersh commented on src/Chapter3.hs:
```diff
@@ -752,6 +830,17 @@ parametrise data types in places where values can be of any general type.
 🕯 HINT: 'Maybe' that some standard types we mentioned above are useful for
   maybe-treasure ;)
 -}
+data Dragon a = Dragon { dragonMagicalPower :: a }
```
Now, after you learned about `newtype`s, you can even make this data type as `newtype` :slightly_smiling_face:

chshersh commented on src/Chapter3.hs:
```diff
@@ -907,9 +996,24 @@ Implement instances of "Append" for the following types:
+instance (Append a) => Append (Maybe a) where
+  append :: Maybe a -> Maybe a -> Maybe a
+  append (Just x1) (Just x2) = Just (append x1 x2)
+  append _ _ = Nothing
```
While this is valid implementation and nothing in the task description says differently, in really it would be more beneficial is this function returned `Just` when at least one argument is `Just`.

chshersh commented on src/Chapter3.hs:
```diff
@@ -970,6 +1074,28 @@ implement the following functions:
+  deriving (Bounded, Eq, Enum)
+
+isWeekend :: DayOfWeek -> Bool
+isWeekend day = day == minBound || day == maxBound
```
Clever ordering of days and usage of standard typeclasses! :brain:

chshersh commented on src/Chapter3.hs:
```diff
@@ -970,6 +1074,28 @@ implement the following functions:
+daysToParty day =
+  let numDays = fromEnum (maxBound :: DayOfWeek) + 1
+      distBetweenDays = (fromEnum Friday - fromEnum day)
+  in distBetweenDays `mod` numDays
```
Very math solution! :1234:

chshersh commented on src/Chapter3.hs:
```diff
@@ -1005,7 +1131,26 @@ properties using typeclasses, but they are different data types in the end.
 contestants, and write a function that decides the outcome of a fight!
 -}
-
+newtype Health' = Health' { getHealth' :: Int }
+newtype Attack' = Attack' { getAttack' :: Int }
+newtype Defense' = Defense' { getDefense' :: Int }
```
Feel free to reuse newtypes from the previous task :slightly_smiling_face: Unfortunately, you can't multiple definitions of the same type within a single module.

chshersh commented on src/Chapter3.hs:
```diff
@@ -1005,7 +1131,26 @@ properties using typeclasses, but they are different data types in the end.
+  , knightDefense' :: Defense'
+  }
+
+class Fighter a where
+  fight' :: a -> b -> String -> (a, b)
+
+doFight :: (Fighter a, Fighter b) => a -> b -> (a, b)
+doFight fighter1 fighter2 = undefined
```
Yes, the error message might not be readable. But what is means is that type variable `b` in the definition of the `fight'` method literally can be anything. So, a valid instance of `Figther` for `Knight'` will look like this:

```haskell
instance Fighter Knight' where
    fight' :: Knight' -> b -> String -> (Knight', b)
```

You see that you have no information about `b` type variable. It can be anything! `Char` or `Bool` or list. You can't fix this type during instance implementation. That's why nothing connects `b` with values the `doFight` function.

You can solve this problem in a multiple ways, but what you can explore is adding a `Fighter` constraint to the `fight'` method:

```haskell
class Fighter a where
  fight' :: Fighter b => a -> b -> String -> (a, b)
```

After that, your instance will look like this:

```haskell
instance Fighter Knight' where
    fight' :: Fighter b => Knight' -> b -> String -> (Knight', b)
```

Now, when implementing this instance, you have access to all the internal implementation details of a `Knight'`, but you also know that type `b` implements the `Fighter` typeclass. You can add more methods to `Fighther` to work with `b` in a generic way.

chshersh reviewed (approved):
Your solutions look amazing!
And so far good progress on the Advanced Task :+1: 

I left some comments and tried to elaborate more on the advanced task. I hope it makes sense :slightly_smiling_face:

jacobprudhomme committed:
Restructured `Castle` type to encode if it has a wall or not

Simplifies some code

jacobprudhomme committed:
Changed `Dragon` type to a newtype

Since it has a single field

jacobprudhomme committed:
Made `append` to only return `Nothing` when both inputs are `Nothing`

jacobprudhomme committed:
Reused newtypes from earlier in the file

jacobprudhomme committed:
Implemented `Fighter` instances for `Monster'` and `Knight'`

Added a bunch of utility methods, without which this task would be impossible

jacobprudhomme committed:
Implemented function to actually do fight

jacobprudhomme committed:
Moved "run" action to `doFight` function

Allows function to exit as-is, without continuing the recursion

jacobprudhomme committed:
Lower-bounded health by 0

jacobprudhomme commented:
@chshersh thanks for the kind comments! You gave me just what I needed to wrap my head around the boss task. As soon as I added the `Fighter b` constraint to the `fight` function, things started working out. My code for the final task is a bit verbose, but I think it should work. If you have the time, I would really appreciate it if you could please take another look at it, to see if it makes sense :smile:

chshersh reviewed (approved):
@jacobprudhomme Your solution makes total sense :+1: Very nice work on the Advanced Task :trophy: 

Only one minor comment: I see that actions in your implementation are lists of `String`. This is not type-safe enough, and ideally they should be custom sum types. But that's just a minor design suggestion :slightly_smiling_face: 

Otherwise, your solutions look great :+1: I think this PR is ready for merge :checkered_flag:

jacobprudhomme commented:
@chshersh great idea! I will add that, thank you

jacobprudhomme committed:
Improved fighting logic

Now, if somebody's health goes to 0 in the middle of one round of attacks, then the battle ends

jacobprudhomme committed:
Made fight commands into a sum type, for better type safety

jacobprudhomme committed:
Took advantage of pattern matching to simplify `doFight`
